File: packages/domain-floorplan/src/opening.ts
import { DomainError } from "./domain-error.js";
import type { OpeningKind } from "./kinds.js";
import { OPENING_KINDS } from "./kinds.js";
import { type LengthMm } from "./length-mm.js";

export interface OpeningProps {
  id: string;
  kind: OpeningKind;
  width: LengthMm;
  height: LengthMm;
  offsetAlongWall: LengthMm;
  sillHeight: LengthMm;
}

export class Opening {
  readonly id: string;
  readonly kind: OpeningKind;
  readonly width: LengthMm;
  readonly height: LengthMm;
  readonly offsetAlongWall: LengthMm;
  readonly sillHeight: LengthMm;

  private constructor(props: OpeningProps) {
    this.id = props.id;
    this.kind = props.kind;
    this.width = props.width;
    this.height = props.height;
    this.offsetAlongWall = props.offsetAlongWall;
    this.sillHeight = props.sillHeight;
  }

  static create(props: OpeningProps): Opening {
    if (!(OPENING_KINDS as readonly string[]).includes(props.kind)) {
      throw new DomainError(
        `Unknown opening kind: ${props.kind}`,
        "INVALID_OPENING",
      );
    }
    if (props.width.value <= 0) {
      throw new DomainError("Opening width must be positive", "INVALID_OPENING");
    }
    if (props.height.value <= 0) {
      throw new DomainError(
        "Opening height must be positive",
        "INVALID_OPENING",
      );
    }
    return new Opening(props);
  }

  withWidth(width: LengthMm): Opening {
    return Opening.create({
      id: this.id,
      kind: this.kind,
      width,
      height: this.height,
      offsetAlongWall: this.offsetAlongWall,
      sillHeight: this.sillHeight,
    });
  }

  occupiesUntilMm(): number {
    return this.offsetAlongWall.value + this.width.value;
  }
}

File: core/src/domain/opening.cpp
#include "toporoom/domain/opening.hpp"

#include "toporoom/domain/error.hpp"

namespace toporoom::domain {

Opening::Opening(OpeningProps props)
    : id_(std::move(props.id)),
      kind_(props.kind),
      width_(props.width),
      height_(props.height),
      offset_along_wall_(props.offset_along_wall),
      sill_height_(props.sill_height) {}

Opening Opening::create(OpeningProps props) {
  if (props.width.value() <= 0) {
    throw DomainError("Opening width must be positive", "INVALID_OPENING");
  }
  if (props.height.value() <= 0) {
    throw DomainError("Opening height must be positive", "INVALID_OPENING");
  }
  return Opening(std::move(props));
}

Opening Opening::with_width(LengthMm width) const {
  return with_placement(width, height_, offset_along_wall_, sill_height_);
}

Opening Opening::with_kind(OpeningKind kind) const {
  OpeningProps props;
  props.id = id_;
  props.kind = kind;
  props.width = width_;
  props.height = height_;
  props.offset_along_wall = offset_along_wall_;
  props.sill_height = sill_height_;
  return Opening::create(std::move(props));
}

Opening Opening::with_placement(LengthMm width, LengthMm height, LengthMm offset_along_wall,
                                LengthMm sill_height) const {
  OpeningProps props;
  props.id = id_;
  props.kind = kind_;
  props.width = width;
  props.height = height;
  props.offset_along_wall = offset_along_wall;
  props.sill_height = sill_height;
  return Opening::create(std::move(props));
}

}  // namespace toporoom::domain
